RFC: Templating `CARGO_TARGET_DIR` to make it the parent of all target directories (#3371)
rust-lang/rfcs: text/3371-cargo-target-dir-templates.md

- Feature Name: `cargo_target_dir_templates`
- Start Date: 2023-01-12
- RFC PR: [rust-lang/rfcs#3371](https://github.com/rust-lang/rfcs/pull/3371)
<!-- - Cargo Issue: [rust-lang/cargo#0000](https://github.com/rust-lang/cargo/issues/0000) -->

# Summary
[summary]: #summary

<!-- One paragraph explanation of the feature. -->

> Introduce templating to `CARGO_TARGET_DIR` to have `cargo` adapts its target directory dynamically depending on (at least) the manifest's path.

Reviewer: typo:
 ```suggestion
Introduce templating to `CARGO_TARGET_DIR` to have `cargo` adapt its target directory dynamically depending on (at least) the manifest's path.
```

Consider expanding the summary somewhat:

> Introduce templating to [`CARGO_TARGET_DIR`](https://doc.rust-lang.org/cargo/reference/config.html#buildtarget-dir).  The value `{manifest-path-hash}` is replaced with a hash based on the path to the current manifest.

Author: Done in https://github.com/rust-lang/rfcs/commit/05b96f8c6638b4bdb72ad240a0e8f8098715f7f6

In this situation, it's much less likely for Cargo to produce invalid state without a `build.rs` deliberately writing outside its target directory.

Two projects can be built in parallel without troubles.

#### Absolute and relative paths

`CARGO_TARGET_DIR` can be either a relative or absolute path, which makes sense since it's mostly intended for a single project, which can then work from its own position to configure the target directory, and that stays the case with templates.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

<!--
This is the technical portion of the RFC. Explain the design in sufficient detail that:

- Its interaction with other features is clear.
- It is reasonably clear how the feature would be implemented.
- Corner cases are dissected by example.

The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
-->

## Templating values of `CARGO_TARGET_DIR`

Templating does not interfere with the resolution order of `CARGO_TARGET_DIR`. From less to most specific:

- Through the `config.toml`:

  ```toml
  [build]

>   target-base-dir = "/absolute/path/to/cache/{manifest-path-hash}"

Author: Thanks! Done in https://github.com/rust-lang/rfcs/commit/226f9ec125f1735c3d1596257d41341aa9e5e815

To prevent collisions by craftings paths, the `<manifest-path-hash>` directory will be computed from a hash of the workspace manifest's full path (and possibly other data, for example `bazel` uses its version and the current user too).

## Symbolic links

In the following situation

```
/Users/
├─ poliorcetics/
│  ├─ projects/
│  │  ├─ actual-crate/
│  │  │  ├─ Cargo.toml
│  │  ├─ symlink-to-crate/ -> actual-crate/
```

When calling `cargo metadata` in the `symlink-to-crate` path, the result contains `"manifest_path": "/Users/poliorcetics/projects/actual-crate/Cargo.toml"` and `"workspace_root":"/Users/poliorcetics/projects/actual-crate"`. This behaviour means that symlinks won't change the final directory used inside `{manifest-path-hash}`, or in other words: symbolic links are resolved.

> ### Handle possibly thousands of directories in a single templated `CARGO_TARGET_DIR` path

Reviewer: I'm not completely convinced that having thousands of directories is likely. That would mean that the host would have 1000s of Cargo projects, without ever cleaning up.

I'd also like to see how this might interact with [Cargo's ongoing GC work](https://blog.rust-lang.org/2023/12/11/cargo-cache-cleaning.html)  covered. Does that help mitigate the concerns?

Reviewer 2: See rust-lang/cargo#13136 which is linked to within the RFC but the link is broken

Reviewer 3: @arlosi Many packaging and build systems have the same pattern. Examples are following, but not limited to: Xcode, MacPorts, FreeBSD ports, yay, Gentoo Portages.

Author: > I'm not completely convinced that having thousands of directories is likely. That would mean that the host would have 1000s of Cargo projects, without ever cleaning up.

Entirely possible in CIs, and even local systems: on my previous machine, I used a common base for all my target directories, with a scheme like `CARGO_TARGET_DIR=/common/path/(basename (pwd))` and just ignored it for my saves. I ended up with more than 250 target dirs from various projects over time (most I uninstalled at some point, the project for garbage collection will be awesome), I can easily see someone reaching the thousands

Its possible values would be:

- `true`: create the symlink and produce an error if it fails
- `"auto"`: create the symlink, produce a warning (or note) but do not fail the command
- `false`: don't create the symlink at all (don't touch it if it exists already though)

### Handling of concurrents builds

It is possible to call `cargo build` for the same project twice with two different target directories to avoid build locks (common when building with different features or to have Rust-Analyzer work in a different target directory for example), which poses a problem for forward links: if `target-dir-link` is active, `cargo` could be replacing the `./target` symlink constantly.

This can be configured through another option, `target-dir-link-replace`, with the following possible values:

- `"notify"` (the default): replace and show the user a note about the replacement
- `"warn"`: replace and show the user a warning about the replacement
- `"silent"`: replace silently
- `"never"`: do not replace

If the forward link doesn't exist already, this option has no effect. If the forward link exist and is the same as the would-be new one, no note nor warning will be produced.

> With both of these options, callers of cargo will be able to use `--config KEY=VALUE` to override it, for example a Rust-Analyzer config could use `cargo.extraArgs = ["--config", "target-dir-link=false"]` to ensure R-A never touches them.

Author: Done in https://github.com/rust-lang/rfcs/commit/2802659393186077d79969e3169926b07b672a4f

### `cargo clean`

Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for the templated version: if it set, `cargo clean` deletes `/path/to/<manifest-path-hash>/` and not `target/`.

# Drawbacks
[drawbacks]: #drawbacks

<!-- Why should we *not* do this? -->

## One more option to find the target directory

This introduces one more option to look at to find the target directory, which may complicate the life of external tools.

This is mitigated by the forward link provided by default by `cargo` when using the templated form of `CARGO_TARGET_DIR`.

## Hitting windows path length limits

Depending on what naming scheme is used (e.g., a very long hash), we could hit the Windows path length limits if not careful.

A mitigation for this is recommending a short prefix (in `CARGO_TARGET_DIR`) and using a hash that doesn't include that many characters but those are only mitigations and do not fully fix the underlying problem.

## Transition period

> During the transition period, any `CARGO_TARGET_DIR` that was defined as containing `{manifest-path-hash}` will change meaning. `cargo`, for at least one stable version of Rust, should provide errors about this and point to either this RFC or its documentation to explain why the incompatiblity arised and how to fix it. In practice, paths with `{` or `}` in it are unlikely, even more with the exact key used by cargo here, so maybe no one will ever see the error, but it's better than silently breaking workflows.

Author: Partially done in https://github.com/rust-lang/rfcs/commit/927226dbd2abd0b6be700b524f3e887ea3841f62

I am unsure about how to formulate in the reference-level explanation without repeating the text of "Drawbacks", should I just move the subsection into the reference section ?

> During the transition period, any `CARGO_TARGET_DIR` that was defined as containing `{manifest-path-hash}` will change meaning. `cargo`, for at least one stable version of Rust, should provide errors about this and point to either this RFC or its documentation to explain why the incompatiblity arised and how to fix it. In practice, paths with `{` or `}` in it are unlikely, even more with the exact key used by cargo here, so maybe no one will ever see the error, but it's better than silently breaking workflows.

Reviewer 2: What we should likely do is move this section entirely to Reference and then just call out the cost of a transition as a bullet point.  No link needed.

> During the transition period, any `CARGO_TARGET_DIR` that was defined as containing `{manifest-path-hash}` will change meaning. `cargo`, for at least one stable version of Rust, should provide errors about this and point to either this RFC or its documentation to explain why the incompatiblity arised and how to fix it. In practice, paths with `{` or `}` in it are unlikely, even more with the exact key used by cargo here, so maybe no one will ever see the error, but it's better than silently breaking workflows.

Author: Done in https://github.com/rust-lang/rfcs/commit/774e32843e841892814a6fd77a060f2cbe5119f0

> ## Brace expansion

Reviewer 2: Wouldn't this deviate from our other "templating" systems in cargo?  For myself, I lean towards consistency but might be convinced otherwise.

> ## Brace expansion

Author: I think that's better answered at the implementation level, once we have some experience with how people use it.

Escaping is always nice, but so is consistency, and I'm really not convinced there is any "non test" path in the world with `{` in it that would appear in a human controlled config for target directories

> Bash has [brace expansion](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html), other shells too. By using `{manifest-path-hash}` we risk users getting bitten by that behaviour. Brace expansion is not activated when there are no `,` nor `..` inside the `{}` so cargo should be fine. Since brace expansion is done at the shell level, cargo won't be able to detect it if it happens.

Reviewer 2: ```suggestion
Bash has [brace expansion](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html), other shells too. By using `{manifest-path-hash}` we risk users getting bitten by that behaviour. Brace expansion is only activated when there are `,` nor `..` inside the `{}` so cargo should be fine. Since brace expansion is done at the shell level, cargo won't be able to detect it if it happens.
```
Removed the double negative to make it easier to parse

tip:  [semantic line feeds](https://rhodesmill.org/brandon/2012/one-sentence-per-line/) make diffing markdown documents easier.

Author: Done in https://github.com/rust-lang/rfcs/commit/3a6fa3e6848a295f7544bf6ac430119b8c1ffc94

- Set `CARGO_TARGET_DIR_REMAP=/home/user/projects=/tmp/cargo-build`
- Compile the crate under `/home/user/projects/foo/` **without** `CARGO_TARGET_DIR` set
- The resulting target directory will be at `/tmp/cargo-build/foo/target`

By making the priority order `CARGO_TARGET_DIR` > `CARGO_TARGET_DIR_REMAP` (when both are absolute paths) we would keep backward compatibility. Or we could disallow having the two set at once, so that they're alternatives and not ordered.

When `CARGO_TARGET_DIR` is relative, the result could be `/tmp/cargo-build/foo/$CARGO_TARGET_DIR`.

Overall, I feel remapping is much harder to implement well and can be added later without interfering with templates in `CARGO_TARGET_DIR` (and without this RFC interfering with remapping), though the design space is probably bigger than the one for this RFC.

## Advantages of using a hash over regular subdirectories

It's possible to achieve most of the same system proposed here by setting a value like `CARGO_TARGET_DIR="/base/dir/{manifest-path-dirs}"`, where a manifest in `/tmp/test1/test2/Cargo.toml` would resolve the build directory to `/base/dir/tmp/test1/test2/`, but most is not *all* of them:

> 1. Predictable depth of subdirectories: even with a hash, depth can be unpredictable in some way since the untemplated part can be as long as necessary but by using a hash of predictable length, users can prepare for the limits imposed by their filesystem (for example on Windows, some forms of path are quite short): it would be surprising for a project to build when in `/` but not in `/tmp/some/what/deep/inside/folders` for example

Reviewer 2: iirc its not depth of subdirectories but the character length of the path.

I'd recommend focusing this on that, e.g.

```
1. Hashes have a fixed length while `manifest-path-dirs` is dependent on the context, making it a hazard for cross-platform compatibility.  Say on Windows the target-dir is rooted in a user tmp dir and the manifest path is inside of the user documents.   Especially combined with corporate policies on names, those base paths alone can take up a good amount of the character budget without getting into project names, etc.
```

Author: Done in https://github.com/rust-lang/rfcs/commit/645d9c41cd581b3c9aaac7f78169cce391ad59dc

> 2. More traversals: by using a hash always cut in the same way to form subdirectories, it's easy to plan traversals for the filesystem

Author: Hopefully clearer with https://github.com/rust-lang/rfcs/commit/0f71cf57ee98bce40695eb39c0c9c12b2d18446b

3. Encourage interactions through `cargo-metadata`: by using paths computed through cargo and not easily derivable from the file tree, future tools will be incentivized to work through `cargo-metadata` to find the target directory, widening adoption and making it easier for the cargo team to ensure nothing breaks in subsequent cargo updates

> 4. Avoid introducing a strong dependency on only the path: by using a hash, cargo can add element to it to help differentiate builds: for example we could use more parameters in the hash, see the relevant section in Future Possibilites

Reviewer 2: These only matter if hashing is being used.  If we give the user a lower level mechanism, then its fine that is predictable and we shouldn't force them to `cargo metadata`.

Author: The subsection is about advantages gained from using a hash so I think those deserve to be here ? They're both advantages and disadvantages to a hash, depending on the use case, but they're certainly relevant I think

- Introduce a form of garbage collection for these target directories so we don't leak them when projects are deleted, see [rust-lang/cargo#13136](https://github.com/rust-lang/cargo#13136) (essentially, backlinks)

## Adding `{xdg-cache-home}`, `{home}`, `{cargo-home}` or `{user-base-target-directory}` as templates

OS-native cache directories are discussed more in details in [rust-lang/cargo#1734](https://github.com/rust-lang/cargo/issues/1734), there are semantic and naming issues to resolve before moving forward with them in cargo (and so as template for `CARGO_TARGET_DIR`).

As a workaround, it is possible to use `CARGO_TARGET_DIR="${XDG_CACHE_HOME:-~/.cache}/cargo-target-directories/{manifest-path-hash}"`.

It won't work in the `config.toml` but it will work with the environment variable and the command line option, both of which override the TOML config.

It is certainly possible to add at least `{home}` and `{cargo-home}` but it can be done in the future without interfering at all with `{manifest-path-hash}`, making it a good option for a future addition without blocking.

## Use templated `CARGO_TARGET_DIR` as the default instead of `target`

> 1. This has several unresolved issues, notably: what default value do we use ? This discussion is already happening in [rust-lang/cargo#1734](https://github.com/rust-lang/cargo/issues/1734)) and does not block this RFC

Reviewer 2: We have a default, what cargo script uses. 

Like I said, this is more of a foot note that moving caches out of `~/.cargo` is a foot note, rather than a concern.

Author: Done in https://github.com/rust-lang/rfcs/commit/d8797514d8b167c09f941b20606a409c35b6db3c

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

<!--
This is the technical portion of the RFC. Explain the design in sufficient detail that:

- Its interaction with other features is clear.
- It is reasonably clear how the feature would be implemented.
- Corner cases are dissected by example.

The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
-->

## Templating values of `CARGO_TARGET_DIR`

Templating does not interfere with the resolution order of `CARGO_TARGET_DIR`. From less to most specific:

- Through the `config.toml`:

  ```toml
  [build]
  target-dir = "/absolute/path/to/cache/{manifest-path-hash}"
  ```

- Through the environment variable: `CARGO_TARGET_DIR="/absolute/path/to/cache/{manifest-path-hash}" cargo build`
- Through the command line flag: `cargo build --target-dir "/absolute/path/to/cache/{manifest-path-hash}"`

> ## Naming
> 
> In the example in the previous section, using a templated `CARGO_TARGET_DIR` with `cargo build` produces named subdirectories. The name of those is computed from the full and canonicalized path to the manifest for the workspace, so building `work-project/crate-1` will still use the directory for the whole workspace during a `cargo build` call.
> 
> This naming scheme is **not considered stable**: the method will probably not change often but `cargo` offers no guarantee and may change it in any release. Tools that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value each time they are invoked.
> 
> The path used for the naming of the final target directory is the one found *after* symlink resolution: `bazel` does it too and I have not found any complaints about this and it has the distinct advantage of allowing to make a symlink to a project somewhere else on the machine (for example to organise work projects) and avoid duplicating the build directory and all its data (which can be quite heavy).
> 

Reviewer 2: I feel like the references to `naming` in here should be changed to refer to "`{manifest-path-hash}`"

```suggestion
## Definition of `{manifest-path-hash}`

In the example in the previous section, `{manifest-path-hash}` was replaced with a relative path. This relative path is computed from the full and canonicalized path to the manifest for the workspace `Cargo.toml`.
By being canonicalized, including resolving of symlinks, symlinked projects will share the same target directory.  This is following the prior art from `bazel` and I have not found any complaints about this.

The hashing and turning the hash into nested directories is **not considered stable**: the method will probably not change often but `cargo` offers no guarantee and may change it in any release. Tools that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value each time they are invoked.
```

Author: Done in 81bac72702e64fe03a0cfaf3050e71d7383922e5

## Impact on `cargo ...` calls

When calling `cargo` with a builtin call (e.g., `build`, `check` or `test`) where a templated `CARGO_TARGET_DIR` is active, `cargo` will first resolve the effective `CARGO_TARGET_DIR` and then proceed with the command as if `CARGO_TARGET_DIR` had been set directly. For third party tools (`cargo-*`), where cargo does not know about the relevant `Cargo.toml`, the tool will have to use [`cargo_metadata`](https://docs.rs/cargo_metadata), as is already expected today, to learn about the effective target directory.

In the same vein, `cargo metadata` fills the target directory information with the absolute path and make no mention of the template in `CARGO_TARGET_DIR` since it can only be used with a single workspace at once.

### `cargo clean`

Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for the templated version: if it set, `cargo clean` deletes `/path/to/<manifest-path-hash>/` and not `target/`.

> ### Providing backlinks

Reviewer 2: If I'm understanding correctly, this RFC specifies no backlink behavior.  If so, this doesn't belong in the Reference section.  Maybe as an Alternative?  We have the GC part of this already in the Future Possibilities.

Author: Done in https://github.com/rust-lang/rfcs/pull/3371/commits/11ab0e38c6351d8d89fa2a6abb063f19c2a18916

# Future possibilities
[future-possibilities]: #future-possibilities

<!--
Think about what the natural extension and evolution of your proposal would
be and how it would affect the language and project as a whole in a holistic
way. Try to use this section as a tool to more fully consider all possible
interactions with the project and language in your proposal.
Also consider how this all fits into the roadmap for the project
and of the relevant sub-team.

This is also a good place to "dump ideas", if they are out of scope for the
RFC you are writing but otherwise related.

If you have tried and cannot think of any future possibilities,
you may simply state that you cannot think of anything.

Note that having something written down in the future-possibilities section
is not a reason to accept the current or a future RFC; such notes should be
in the section on motivation or rationale in this or subsequent RFCs.
The section merely provides additional information.
-->

- Introduce remapping into the concept in some way.

> - Introduce a form of garbage collection. Expanded upon this [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/last-use.20tracking).

Reviewer 2: ```suggestion
- Introduce a form of garbage collection for these target directories so we don't leak them when projects are deleted, see rust-lang/cargo#13136
```
- Updating of link
- Clarifying why this is relevant

Author: Done in https://github.com/rust-lang/rfcs/pull/3371/commits/2ae536e3d73ae211b0342aa0448f95123a240516

-->

For a single project, it is possible to use the `CARGO_TARGET_DIR` environment variable (or the `target-dir` TOML config option or the `--target-dir` command-line flag) with either an absolute or relative path to change the position of the `target/` directory used for build artifacts during compilation with Cargo.

While this option is useful for single-project environments (simple CI builds, builds through other build systems like Meson or Bazel), in multi-projects environment, like personal machines or repos with multiple workspaces, it conflates every build directory under the configured path: `CARGO_TARGET_DIR` directly replaces the `<workspace>/target/` directory.

Templating introduces one new templating key for `CARGO_TARGET_DIR`, in the same spirit as [the index configuration format][icf]:

- `{manifest-path-hash}`: a hash of the manifest's absolute path as a path. This is **not** an absolute path.

It can be used like this: `CARGO_TARGET_DIR="$HOME/.cache/cargo-target-dirs/{manifest-path-hash}"`.

> When compiling `/home/ferris/src/cargo/` with user `ferris`, `manifest-path-hash` would be something like `ab/cd/<rest of hash>` and the artifacts would be found in `/home/ferris/.cache/cargo-target-dirs/ab/cd/<rest of hash>/...`. Note the hash used and the path derived from that for `{manifest-path-hash}` are implementation details and the values here are just example.

Reviewer 4: ```suggestion
When compiling `/home/ferris/src/cargo/` with user `ferris`, `manifest-path-hash` would be something like `ab/cd/<rest of hash>` and the artifacts would be found in `/home/ferris/.cache/cargo-target-dirs/ab/cd/<rest of hash>/...`. Note the hash used and the path derived from that for `{manifest-path-hash}` are implementation details and the values here are just an example.
```

Author: Done in 0a38c0767eeae0d6f0bdb433c9207c167e4f3245

1. Hashes have a fixed length while `manifest-path-dirs` is dependent on the context, making it a hazard for cross-platform compatibility. Say on Windows the target-dir is rooted in a user tmp dir and the manifest path is inside of the user documents. Especially combined with corporate policies on names, those base paths alone can take up a good amount of the character budget without getting into project names, etc.
2. Encourage interactions through `cargo-metadata`: by using paths computed through cargo and not easily derivable from the file tree, future tools will be incentivized to work through `cargo-metadata` to find the target directory, widening adoption and making it easier for the cargo team to ensure nothing breaks in subsequent cargo updates
3. Avoid introducing a strong dependency on only the path: by using a hash, cargo can add element to it to help differentiate builds: for example we could use more parameters in the hash, see the relevant section in Future Possibilites

## Alternatives to the transition period

While unlikely, the transition period may break workflows by introducing errors for previously valid target directories.

> Several alternatives exist for this:

Reviewer 4: It would be worth talking about what happens if the config option is turned on and off, or if older and newer Cargo versions are invoked with e.g. `cargo +version build`. And to consider the behavior both with and without forward links. For example, let's say Cargo version 1.500 supports templating:

* Fresh build with Cargo 1.500
* Fresh build with 1.500 and then with Rust 1.499
* Fresh build with 1.499 and then with Rust 1.500
* `cargo +1.499 clean`

Old Rust versions will be around for a while, and typical user behavior would be to set it in the global env or .cargo/config, and then forget about it.

For comparison, targo needs to be invoked once to create the symlink, and then most cargo commands work fine as usual regardless of version.

Author: > It would be worth being explicit

On the contrary, I don't think it's worth it: that would be better determined by experimentation I think, and once we have something of a consensus by testers we can finalize and write a definite table for it.

There are effectively two options:

- Templating
- Forwarding

The second does not matter if the first is inactive.

If templating is active it means `CARGO_TARGET_DIR` is set in some way, in which case the behaviour is known: `cargo` creates a directory as directed, so `CARGO_TARGET_DIR=/tmp/{target}` would create exactly `/tmp/{target}`, as it does today already.

I have a few rules I would like to see because they're also how `cargo` operates today (for the non-templating rules):

- No `CARGO_TARGET_DIR` setting: work with `./target`, regardless of if it's a symlink or directory, as is done today
- Using `CARGO_TARGET_DIR=/no-template`, ignore `./target` completely, as is done today
- Using `CARGO_TARGET_DIR=/tmp/{manifest-path-hash}`
  - `1.499`: ignore `./target` completely, as is done today, works in exactly `/tmp/{manifest-path-hash}`
  - `1.500`: works in `/tmp/<hash path>`
    - Forwarding:
      - `./target` is symlink: replace it, work with it
      - `./target` is not symlink: ignore it completely
    - No forwarding: ignore `./target` completely

Subsequent commands where the target dir (regardless of position) already exist would just work with it, as is done today: I can build with 1.76 and 1.77 in the same repo, it will just work in the same target dir and do whatever it wants with the artifacts and such.

But all of those rules (barring those that are already in action in released versions of `cargo`) shouldn't be set in stone IMO, I would prefer having users coming to us to tell us what they (didn't) like about them during the unstable period.

If enough people want to see them in the RFC then I can them though, since it will likely mean there is some consensus on them already

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

<!--
This is the technical portion of the RFC. Explain the design in sufficient detail that:

- Its interaction with other features is clear.
- It is reasonably clear how the feature would be implemented.
- Corner cases are dissected by example.

The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
-->

## Templating values of `CARGO_TARGET_DIR`

Templating does not interfere with the resolution order of `CARGO_TARGET_DIR`. From less to most specific:

- Through the `config.toml`:

  ```toml
  [build]
  target-dir = "/absolute/path/to/cache/{manifest-path-hash}"
  ```

- Through the environment variable: `CARGO_TARGET_DIR="/absolute/path/to/cache/{manifest-path-hash}" cargo build`
- Through the command line flag: `cargo build --target-dir "/absolute/path/to/cache/{manifest-path-hash}"`

## Definition of `{manifest-path-hash}`

> In the example in the previous section, `{manifest-path-hash}` was replaced with a relative path. This relative path is computed from the full and canonicalized path to the manifest for the workspace `Cargo.toml` (or the `script.rs` file directly for cargo-scripts).

Reviewer 5: This will be the very first templating supported in `.cargo/config.toml`. Is the syntax flexible enough to extend to other configuration like [`[build.rustflags]`](https://doc.rust-lang.org/nightly/cargo/reference/config.html#buildrustflags), [`[env]`](https://doc.rust-lang.org/nightly/cargo/reference/config.html#env). I can see it useful in other issues like

* https://github.com/rust-lang/cargo/issues/12193
* https://github.com/rust-lang/cargo/issues/7352

Author: I see no reason for it to be limited.

`{name}` is not used anywhere currently IIRC and the most likely places it can crop up unexpectedly are paths, where this RFC already discuss how unlikely that is.

I suppose anything where cargo shells out to another program can also have it come up, but I'm not sure we can even design for every possible `target.TTT.runner` in eternity

Reviewer 5: Nice. Thanks for the reply. Maybe it is worth putting this in future posibilities in some way?

Author: Done in https://github.com/rust-lang/rfcs/pull/3371/commits/d549cb12889d5a2c4a4dd64b5a8a0e3ee7b49aa0 :) I tried noting possible future problems I could find in 10 minutes of thinking and why they're not limiting

- Do we want to add forward links for untemplated `CARGO_TARGET_DIR` too in the process of this RFC ?
- Do we want to differentiate according to users ? `bazel` is a generic build tool, whereas `cargo` is not, so maybe differentiating on users is not necessary for the latter ?

> ## Providing forward links

Author: Done in https://github.com/rust-lang/rfcs/pull/3371/commits/6540d11b8e2a2f4def265989a925932843ced8f3

I expanded a lot on them, it's one of the part that changed the most, I think I have a good base but it introduces 2 new options so I don't know if it's too much
